### ISSUE TITLE
Update to compile under recent Haskell Platform

### ISSUE THREAD

**Issue by RayZopf** [raysilent/lslforge#20]
https://code.google.com/archive/p/lslforge/issues/24  
Posted on Aug 22, 2012 by Quick Kangaroo
What steps will reproduce the problem? 1. Install Haskell-platform >=11.0 2. cabal install

What is the expected output? What do you see instead? Expected: completed build of Haskell LSLForge Actual: irreconcilable package dependencies, varying depending on specific installation

What version of the product are you using? On what operating system? current git head (0.1.2)

Please provide any additional information below. As far as I can tell, it is difficult (impossible?) to compile on recent OSX, as the suggested ghc is 6.10.4. But versions prior to 6.12 had issues on OSX with generating 32bit executables. More recent versions do not support the older versions of dependencies required by LSLForge, for instance: haskell98-1.1.0.1 requires array ==0.3.\* but LSLForge requires array ==0.2.*

**Comment by simon-nicholls** [comment]
This seems like a worthy quest, and I've made a start.

Currently I have a fork that works with ghc-6.12.3, and I'll be hooking up the tests into a proper Cabal test-suite before investigating a Monad Transformer issue. Once that is resolved, I'll be working on ghc-7.0.* changes.

**Comment by simon-nicholls** [comment]
OK, my ghc-7.0.4 branch is now merged to master.

Travis CI is building and testing LSLForge successfully for all but a few `CompilationServer` unit tests, and those don't actually look to be in sync with the source.

A quick test with Eclipse shows the LSLForge binary to be functional, but there are a bunch of `fail` calls throughout the codebase, and these cause failures with transformers >= 0.2.2.0.

It can be built with the latest `cabal-install` binary download, on Linux and macOS at least, so the build story is much nicer now, but there's a long way to go yet.

**Comment by simon-nicholls** [comment]
Now at ghc-8.0.2

**Comment by raysilent** [comment]
Awesome news, @simon-nicholls !

**Comment by simon-nicholls** [comment]
Over to you for review! It would be good if you could trial my pull request against some production LSL code before merging.

**Comment by raysilent** [comment]
Agreed. It might be a good idea to compare compiled LSL-files for starters.

**Comment by raysilent** [comment]
Simon, we have some issues with the Java generated code from Haskell. Hope you see the message

**Comment by raysilent** [comment]
@simon-nicholls, I'm about to close this, but would appreciate if you look at the recent changes at the https://github.com/raysilent/lslforge/tree/0.1.9.4 tree. I had to introduce a few workarounds that resulted from the code that you upgraded to the latest Haskell libraries. Please have a look. Thank you

**Comment by simon-nicholls** [comment]
Hi,

Sorry for being unresponsive, but things are so rough that I can’t even work right now.

The SerializationGenerator change looks on-track to me, and the TupleX stuff will have to be fixed at some point - seems reasonable if it works!

Cheers,
Simon

**Comment by raysilent** [comment]
No problem @simon-nicholls. You did such a fantastic job, that wasn't even possible to imagine that would ever be accomplished.

After testing for quite a while, with the silly patch I provided, it seems working, although for some reason I noticed that that "TupleX" initialization has to happen earlier in the generated/InitAll.java, so I'm manually moving 2 lines up at the moment. The symptom is that the whole Eclipse hangs during initialization, the window freezes and doesn't want to continue. LSLForge executable, even if stopped, restarts itself and nothing really happens. I wonder if this *is* related or not, but this is my observation. 

So first of all, they have to be there (which was broken) and probably have to be moved (my patch basically appends them). If it all makes sense to you, I can't figure out why TupleX disappeared in the first place.
